core/src/main/scala/kafka/server/RaftRequestHandler.scala

```diff
@@ -48,6 +59,54 @@ class RaftRequestHandler(networkChannel: KafkaNetworkChannel,
             requestBody,
             response => sendResponse(request, Some(response)))
 
+        case ApiKeys.API_VERSIONS =>
+          sendResponse(request, Option(ApiVersionsResponse.apiVersionsResponse(10, 2)))
+
+        case ApiKeys.METADATA =>
+          val topics = new MetadataResponseData.MetadataResponseTopicCollection
+          if (connectionState.hostInfo().isPresent) {
+            topics.add(new MetadataResponseTopic()
+              .setErrorCode(Errors.NONE.code)
+              .setName("__cluster_metadata")
+              .setIsInternal(true)
+              .setPartitions(Collections.singletonList(new MetadataResponsePartition()
+                .setErrorCode(Errors.NONE.code())
+                .setPartitionIndex(0)
+                .setLeaderId(connectionState.id().toInt))))
+          }
+
+          val brokers = new MetadataResponseData.MetadataResponseBrokerCollection
+
+          if (connectionState.hostInfo().isPresent) {
+            val socketAddress = connectionState.hostInfo().get().address
+            brokers.add(new MetadataResponseData.MetadataResponseBroker()
+              .setNodeId(connectionState.id().toInt)
+              .setHost(socketAddress.getHostName)
+              .setPort(socketAddress.getPort))
+          }
+
+          sendResponse(request, Option(new MetadataResponse(
+            new MetadataResponseData()
+              .setTopics(topics)
+              .setBrokers(brokers))))
+
+        case ApiKeys.PRODUCE =>
+          val produceRequest = request.body[ProduceRequest]
+          counter.incrementByRecords(
```

Comment:
Hmm, `incrementByRecords` is a strange API. The payloads sent through the produce API are arbitrary. I think we should just use a separate state machine which has an API to write data, but has no-op implementations for `accept` and `apply`.